src/main/webapp/app/exercises/shared/result/result.service.ts

```diff
@@ -31,12 +36,64 @@ export class ResultService implements IResultService {
     private participationResourceUrl = SERVER_API_URL + 'api/participations';
 
-    constructor(private http: HttpClient) {}
+    constructor(private http: HttpClient, private translateService: TranslateService) {}
 
     find(resultId: number): Observable<EntityResponseType> {
         return this.http.get<Result>(`${this.resultResourceUrl}/${resultId}`, { observe: 'response' }).pipe(map((res: EntityResponseType) => this.convertDateFromServer(res)));
     }
 
+    getResultString(result: Result, exercise: Exercise): string {
+        const relativeScore = roundValueSpecifiedByCourseSettings(result.score!, getCourseFromExercise(exercise));
+        const points = roundValueSpecifiedByCourseSettings((result.score! * exercise.maxPoints!) / 100, getCourseFromExercise(exercise));
+        if (exercise.type === ExerciseType.PROGRAMMING) {
+            return this.getResultStringProgrammingExercise(result, exercise as ProgrammingExercise, relativeScore, points);
+        } else {
+            return this.translateService.instant(`artemisApp.result.resultStringNonProgramming`, {
+                relativeScore,
+                points,
+                maxPoints: exercise.maxPoints,
+            });
+        }
+    }
+
+    private getResultStringProgrammingExercise(result: Result, exercise: ProgrammingExercise, relativeScore: number, points: number): string {
+        let buildAndTestMessage: string;
+        if (result.submission && (result.submission as ProgrammingSubmission).buildFailed) {
+            buildAndTestMessage = this.translateService.instant('artemisApp.result.resultStringBuildFailed');
+        } else if (!result.testCaseCount || result.testCaseCount === 0) {
+            buildAndTestMessage = this.translateService.instant('artemisApp.result.resultStringBuildSuccessfulNoTests');
+        } else {
+            buildAndTestMessage = this.translateService.instant('artemisApp.result.resultStringBuildSuccessfulTests', {
+                numberOfTestsPassed: result.passedTestCaseCount === 255 ? '255+' : result.passedTestCaseCount,
+                numberOfTestsTotal: result.testCaseCount === 255 ? '255+' : result.testCaseCount,
+            });
+        }
+
+        let resultString: string;
+        if (result.codeIssueCount && result.codeIssueCount > 0) {
+            resultString = this.translateService.instant('artemisApp.result.resultStringProgrammingCodeIssues', {
+                relativeScore,
+                buildAndTestMessage,
+                numberOfIssues: result.codeIssueCount === 255 ? '255+' : result.codeIssueCount,
+                points,
+                maxPoints: exercise.maxPoints,
+            });
+        } else {
+            resultString = this.translateService.instant(`artemisApp.result.resultStringProgramming`, {
+                relativeScore,
+                buildAndTestMessage,
+                points,
+                maxPoints: exercise.maxPoints,
+            });
+        }
```

Comment:
```suggestion
        let resultString = getBaseResultStringProgrammingExercise(…);
```
This if/else can be moved to a separate method.